### ISSUE TITLE
Copy a file (content) to one of X clipboards

### ISSUE THREAD

**Issue by tigerjack** [ranger/ranger#959]
#### ISSUE TYPE
- Improvement/feature request

#### RUNTIME ENVIRONMENT

- Operating system and version: Arch Linux 4.13.5-1-ARCH
- ranger version: ranger-master 1.9.0b5
- Python version: 3.6.2 (default, Jul 20 2017, 03:52:27) [GCC 7.1.1 20170630]
- Locale: en_GB.UTF-8

#### EXPECTED BEHAVIOR
Ideally, it could be of great use to copy a file content (not necessarily text, but also binary data) from a (many) file(s) selected in ranger to the X clipboard. In such a way, you could for example attach a file in your mail program, insert an image into a word processor or copy a file into a web page (Github f.e. support the direct paste of files from the clipboard).

#### CURRENT BEHAVIOR
There is nothing like that at the moment.

**Comment by chrissound** [comment]
You could easily do this with xclip and a keybinding.

**Comment by tigerjack** [comment]
@chrissound Well, I don't think it's too easy. In order to copy a binary file (f.e. an image) with xclip (or even CopyQ) you have to specify the correct mime type of the file.
I've also played around with some bash script, but it doesn't always work.

If it could be useful, here is the content of the script with the different alternatives I've tried.

```
#!/usr/bin/sh
t=$(file -b --mime-type "$1")
#cat "$1" | xclip -selection c -t "$t"
#xclip -selection c -t "$t" -i "$1"
copyq write $t - < $1 && copyq select 0

```

**Comment by chrissound** [comment]
Oh right - I have found that you can determine the mimetype with `file --mime-type filename.txt` for example.

**Comment by tigerjack** [comment]
@chrissound yeah, it's what I've done in the script. However, it doesn't work with all files and applications. I mean, I tried f.e. to use the script above with a png file and I was able to correctly paste the file to a Telegram chat or a github coment.
However, when I tried different file (pdf, m4a, webm, ...) the script doesn't work, in the sense that the file is copied to the clipboard, but it can't be pasted for whatever reason.

**Comment by qoxxxx** [comment]
Is `xclip-copyfile`, `xclip-cutfile` and `xclip-pastefile` what you're looking for ? You'd have to be willing to bind `xclip-pastefile` to something (like ctrl-meta-f).

**Comment by tigerjack** [comment]
@qoxxxx Unfortunately, it doesn't work. I already tried that option and it can be useful to copy/paste files from a directory to another in a terminal (even a remote one) but not from the terminal to a gui application.

**Comment by qoxxxx** [comment]
Oh ok, it's working for me with Thunar for example, but I didn't try others.

**Comment by tigerjack** [comment]
@qoxxxx what do you mean by it's working with Thunar? You can copy files from thunar and paste them to another X application?

**Comment by chrissound** [comment]
Regardless - I don't think this functionality should be bundled with ranger. And instead should just rely on a keybinding.

Any issues would ideally be handled xclip or it's alternatives.

**Comment by tylerlmz1** [comment]
for noobs like me that want a step by step instructions on how to do this:

### Install xclip

`sudo apt install xclip`

### put this line in your rc.conf

`map yc shell cat %p | xclip -sel clip`

it binds `yc` to copying the file's content

### use it
press yc on a file

( yc stands for yank content )

**Comment by Liguobing00** [comment]
As suggested by tylerlmz1, I put the following lines in rc.conf to copy png files:

**# copy png to clipboard
map yc shell xclip -sel clip -t image/png %p**

**Comment by jajpater** [comment]
I combined the answer @qoxxxx and @tylerlmz1 

    map yc shell xclip-copyfile %p 

This  works on all filetypes and even with a list of files. 
It doesn't work if a filename has spaces though.

**Comment by tylerlmz1** [comment]
> I combined the answer @qoxxxx and @tylerlmz1 
> 
>     map yc shell xclip-copyfile %p 
> 
> This  works on all filetypes and even with a list of files. 
> It doesn't work if a filename has spaces though. 

Would it work with filenames with spaces if you put double quotes around %p ?

**Comment by toonn** [comment]
It should actually work fine with filenames containing spaces. The expansion of the macros should be properly escaped for `shell`.

**Comment by jajpater** [comment]
I seem to be in error. It works with files with spaces in the names even without quotes!

**Comment by edjubert** [comment]
@toonn real easy indeed:
```bash
map yc shell cat %p | xclip -sel clip -t `file -b --mime-type %p` %p
```

**Comment by MrIvanPlays** [comment]
help. I tried both
```
map yc shell xclip-copyfile %p 
```
and
```
map yc shell xclip -sel clipboard -t `file -b --mime-type %p` %p
```
and they didn't work for me. I tried copying a jar so I can paste it into discord (to send it).

**Comment by scottkosty** [comment]
@MrIvanPlays are you using Wayland by chance?

What do you get on the command line? For example, does the following work?

    xclip -sel clipboard <filename>

**Comment by scottkosty** [comment]
> I'm not using wayland.

Good to know.

> `xclip -sel clipboard <filename>` copied a jpg file with the clipboard being set with some random UTF-8 characters when I tried pasting it.

First test with text files. Does that work? If so, does it work in ranger? Let's first debug that (more simple) case before trying other mime types.

**Comment by MrIvanPlays** [comment]
> First test with text files. Does that work? If so, does it work in ranger? Let's first debug that (more simple) case before trying other mime types.

Running the command from console with a text file sent the text file's contents to my clipboard and it had the same behavior inside ranger with me running `:script xclip -sel clipboard %p` with the text file selected.

**Comment by scottkosty** [comment]
Does it work when you add the `map` command and try the keyboard shortcut? Let's keep testing this on a simple text file before others.

**Comment by MrIvanPlays** [comment]
> Does it work when you add the `map` command and try the keyboard shortcut? Let's keep testing this on a simple text file before others.

worked fine. now what about other file types?

**Comment by scottkosty** [comment]
To test on the command-line, do the following:

    xclip -sel clipboard -t `file -b --mime-type yourfile` yourfile

If that works, it should work from ranger as well.

**Comment by scottkosty** [comment]
> that did not work from the command line

Good to know. It is helpful if you give more details cases like these. For example, instead of saying "did not work", do you meant that you received an error? If so, what was the error? Or perhaps there was no error but nothing was copied to the clipboard? Or perhaps something was copied to the clipboard but the mime was not set correctly?

Are you running a clipboard manager? For example, I use [https://github.com/hluk/CopyQ](> that did not work from the command line

Good to know. It is helpful if you give more details cases like these. For example, instead of saying "did not work", do you meant that you received an error? If so, what was the error? Or perhaps there was no error but nothing was copied to the clipboard? Or perhaps something was copied to the clipboard but the mime was not set correctly?

Are you running a clipboard manager? For example, I use [CopyQ](url).

By the way, it's good, in some sense, that it doesn't work from the command line. It would be a bit harder to debug if it worked on the command line but didn't work in ranger.)

By the way, it's good, in some sense, that it doesn't work from the command line. It would be a bit harder to debug if it worked on the command line but didn't work in ranger.

**Comment by MrIvanPlays** [comment]
no errors, nothing sent to the clipboard. I am using some clipboard manager but I don't know it's name. I am using a custom config ~~I totally did not copy~~ for BSPWM. 

I'm not entirely sure if its a clipboard manager either. I have this in my application icons on my polybar: 
![image](https://user-images.githubusercontent.com/37958105/200905998-5fee957e-0950-4abf-9cab-f57a40eb280a.png)
which lets me switch between multiple clipboards.

**Comment by scottkosty** [comment]
What is the output of

    file -b --mime-type yourfile

for the file you're testing with?

**Comment by MrIvanPlays** [comment]
> What is the output of
> 
> ```
> file -b --mime-type yourfile
> ```
> 
> for the file you're testing with?

a jar file, the output of this is `application/zip`.

**Comment by MrIvanPlays** [comment]
turns out im using parcellite clipboard manager.

**Comment by MrIvanPlays** [comment]
I've migrated to copyq but i still cannot get a jar file or any other mime type except text to be copied. I've also tried the copyq command about copying and still doesn't want to send the jar to my clipboard.

**Comment by scottkosty** [comment]
I see, so the issue is that the file is compressed. I've only tried the copying with text mime types and picture mime types. Otherwise, if I just copy the path of the file (`yp` in ranger), and paste the path wherever I need to attach it.

I'm not sure what you want to do is a system-specific issue. I don't think the `yc` script is meant to work in this case, but that's just my guess.

**Comment by MrIvanPlays** [comment]
copying the path and appending `file://` to it seems to work fine with any mime type. thanks for the help though.

**Comment by stringinging** [comment]
under GNOME desktop, you can try this
`map yc shell echo -e "copy\nfile://"%p"\0" | xclip -i -selection clipboard -t x-special/gnome-copied-files`
